### PR TITLE
Implement Texture of Memories [LC]

### PR DESCRIPTION
**Pull request by mnpqraven**

### CODE REVIEW

File: internal/lightcone/preservation/textureofmemories/textureofmemories.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+// wearer doesn't have Shield after attacked
+func onAfterBeingAttacked(mod *modifier.ModifierInstance, e event.AttackEndEvent) {
+	isShielded := mod.Engine().IsShielded(mod.Owner())
+	isOnCd := mod.Engine().HasModifier(mod.Owner(), modcd)
+
+	// no shield + lc effect is off-cd
+	if !isShielded && !isOnCd {
+		state := mod.State().(State)
+		// apply shield
+		mod.Engine().AddShield(modshield, info.Shield{
```
Comment (unleashurgeek) [inline internal/lightcone/preservation/textureofmemories/textureofmemories.go:79]: this shield is only supposed to last 2 turns. Want to do the following:
* Create a shield modifier that `OnAdd` calls `AddShield` like you currently do
* Have the shield modifier implement the `OnRemove` listener and call `RemoveShield` in it
* Set this modifier's `StatusType` to `BUFF`
* Here call `AddModifier` with its `Duration` set to 2 turns

File: internal/lightcone/preservation/textureofmemories/textureofmemories.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+			Source:     mod.Owner(),
+			Target:     mod.Owner(),
+			BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_MAX_HP: state.shieldAmt},
+		})
+
+		// apply cd modifier
+		mod.Engine().AddModifier(mod.Owner(), info.Modifier{
+			Name:     modcd,
+			Source:   mod.Owner(),
+			Duration: 2,
```
Comment (unleashurgeek) [inline internal/lightcone/preservation/textureofmemories/textureofmemories.go:89]: CD should be 3 turns

File: internal/lightcone/preservation/textureofmemories/textureofmemories.go [hunk 3]
```diff
@@ -74,19 +83,33 @@ func onAfterBeingAttacked(mod *modifier.ModifierInstance, e event.AttackEndEvent
...
 	// no shield + lc effect is off-cd
 	if !isShielded && !isOnCd {
-		state := mod.State().(State)
-		// apply shield
-		mod.Engine().AddShield(modshield, info.Shield{
-			Source:     mod.Owner(),
-			Target:     mod.Owner(),
-			BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_MAX_HP: state.shieldAmt},
+		// apply shield as buff with 2t duration
+		mod.Engine().AddModifier(mod.Owner(), info.Modifier{
```
Comment (unleashurgeek) [inline internal/lightcone/preservation/textureofmemories/textureofmemories.go:87]: the OnAdd listener of shield buff uses `State`, so you need to pass a state into this modifier

File: internal/lightcone/preservation/textureofmemories/textureofmemories.go [hunk 4]
```diff
@@ -87,6 +87,7 @@ func onAfterBeingAttacked(mod *modifier.ModifierInstance, e event.AttackEndEvent
 		mod.Engine().AddModifier(mod.Owner(), info.Modifier{
 			Name:     modshieldbuff,
 			Source:   mod.Owner(),
+			State:    mod.State().(State),
```
Comment (mnpqraven) [inline internal/lightcone/preservation/textureofmemories/textureofmemories.go:90]: would passing state down like this work ? For `modshieldbuff`'s `onAdd` to get the required state, i give the adder function the current state in `mod`'s `onAfterBeingAttacked` like passing props in js

**Review by unleashurgeek — APPROVED** [review]

**Pull request merged**